cmd/containerboot/kube.go

```diff
@@ -125,3 +132,62 @@ func (kc *kubeClient) storeCapVerUID(ctx context.Context, podUID string) error {
 	}
 	return kc.StrategicMergePatchSecret(ctx, kc.stateSecret, s, "tailscale-container")
 }
+
+// waitForConsistentState waits for tailscaled to finish writing state if it
+// looks like it's started. It is designed to reduce the likelihood that
+// tailscaled gets shut down in the window between authenticating to control
+// and finishing writing state. However, it's not bullet proof because we can't
+// atomically authenticate and write state.
+func (kc *kubeClient) waitForConsistentState(ctx context.Context) error {
+	var logged bool
+
+	bo := backoff.NewBackoff("", logger.Discard, 2*time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		secret, err := kc.GetSecret(ctx, kc.stateSecret)
```

Comment:
@tomhjp altough we originally talked about checking for state consistency this way, I am getting concerned about the large number of kube API server calls. It seems like in the worst case scenario, with the backoff that we have, we can end up with a bit over a 100
I gave this a bit more thought and I think we could achieve the same if here we just call https://github.com/tailscale/tailscale/blob/c2761162a002e65e2305f7570b2c54c561ac151f/client/tailscale/localclient.go#L1438 this method instead and check that ID is set
I've experimented with this branch and can see that too solving #14080, but without any API server calls (as we now read kube state from memory that's populated by writes).
WDYT?

Reply:
I'm not able to reproduce the 100 API calls figure you mention; I calculated the max number of calls as ~21 (depending on jitter) and a quick unit test bore out the same result, which (if correct) seems pretty reasonable as a maximum over a period of 25 seconds. The median and mode should comfortably be 1, as the common case will be that we don't hit the window where we have a race.

I also had a look at the code paths involved in writing state vs setting current profile in the local backend, and I think relying on the local backend makes the chance for a race slightly greater because in tailscaled we update the profile manager's current profile just before we write to our state store (in particular, look at the write of _current-profile which is the last state key to get written), and that write could take arbitrarily long if the kube API server is overloaded, which is exactly one class of scenario where we might expect the race window to be larger.

On balance, I'm leaning towards leaving this as-is, because the kube Secret represents the source of truth for the state next time the Pod is going to start up, and I'd rather consult the source of truth if the load isn't excessive. As a best of both worlds, we could poll the local API first and then poll the kube Secret, but my initial sense is that that would feel like premature optimisation.